src/connector/src/source/kafka/mod.rs

```diff
@@ -58,127 +56,21 @@ pub struct KafkaProperties {
     #[serde(rename = "properties.group.id", alias = "kafka.consumer.group")]
     pub consumer_group: Option<String>,
 
-    /// Security protocol used for RisingWave to communicate with Kafka brokers. Could be
-    /// PLAINTEXT, SSL, SASL_PLAINTEXT or SASL_SSL.
-    #[serde(rename = "properties.security.protocol")]
-    security_protocol: Option<String>,
-
-    // For the properties below, please refer to [librdkafka](https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md) for more information.
-    /// Path to CA certificate file for verifying the broker's key.
-    #[serde(rename = "properties.ssl.ca.location")]
-    ssl_ca_location: Option<String>,
-
-    /// Path to client's certificate file (PEM).
-    #[serde(rename = "properties.ssl.certificate.location")]
-    ssl_certificate_location: Option<String>,
-
-    /// Path to client's private key file (PEM).
-    #[serde(rename = "properties.ssl.key.location")]
-    ssl_key_location: Option<String>,
-
-    /// Passphrase of client's private key.
-    #[serde(rename = "properties.ssl.key.password")]
-    ssl_key_password: Option<String>,
-
-    /// SASL mechanism if SASL is enabled. Currently support PLAIN, SCRAM and GSSAPI.
-    #[serde(rename = "properties.sasl.mechanism")]
-    sasl_mechanism: Option<String>,
-
-    /// SASL username for SASL/PLAIN and SASL/SCRAM.
-    #[serde(rename = "properties.sasl.username")]
-    sasl_username: Option<String>,
-
-    /// SASL password for SASL/PLAIN and SASL/SCRAM.
-    #[serde(rename = "properties.sasl.password")]
-    sasl_password: Option<String>,
+    pub common: Option<KafkaCommon>,
 
-    /// Kafka server's Kerberos principal name under SASL/GSSAPI, not including /hostname@REALM.
-    #[serde(rename = "properties.sasl.kerberos.service.name")]
-    sasl_kerberos_service_name: Option<String>,
-
-    /// Path to client's Kerberos keytab file under SASL/GSSAPI.
-    #[serde(rename = "properties.sasl.kerberos.keytab")]
-    sasl_kerberos_keytab: Option<String>,
-
-    /// Client's Kerberos principal name under SASL/GSSAPI.
-    #[serde(rename = "properties.sasl.kerberos.principal")]
-    sasl_kerberos_principal: Option<String>,
-
-    /// Shell command to refresh or acquire the client's Kerberos ticket under SASL/GSSAPI.
-    #[serde(rename = "properties.sasl.kerberos.kinit.cmd")]
-    sasl_kerberos_kinit_cmd: Option<String>,
-
-    /// Minimum time in milliseconds between key refresh attempts under SASL/GSSAPI.
-    #[serde(rename = "properties.sasl.kerberos.min.time.before.relogin")]
-    sasl_kerberos_min_time_before_relogin: Option<String>,
-
-    /// Configurations for SASL/OAUTHBEARER.
-    #[serde(rename = "properties.sasl.oauthbearer.config")]
-    sasl_oathbearer_config: Option<String>,
+    #[serde(flatten)]
+    extra: Option<HashMap<String, Value>>,
```

Comment:
It seems that `extra` is only used in `extract_common`, and could be eliminated if we directly parse into `common`. 🤔

Reply:
That's not how it works. All properties are passed in a single-layer hashmap and cannot be parsed into a nested struct unless we impl deserialize trait for each case.

Reply:
It's still weird to have both `extra` and `common` in a single struct, where they are actually the same stuff. If there's no better way to resolve this, better add some comments.

Reply:
for example, the json is 

```
{
"kafka.brokers": "xxx",
"topic": "xxx",
"scan.startup.mode": "earliest"
}
```

serde_json will only get `scan_startup_mode` for `KafkaProperties` since the name matches its member but ignores other fields. 
if the json itself is nested, your solution will work 

```
{
"common": {
  "kafka.brokers": "xxx",
  "topic": "xxx"
}
"scan.startup.mode": "earliest"
}
```

Reply:
BTW, will `#[serde(flatten)]` on `common` work? [ref here](https://serde.rs/attr-flatten.html)

Reply:
> BTW, will `#[serde(flatten)]` on `common` work? [ref here](https://serde.rs/attr-flatten.html)

I wrote an example, it works. I suggest to use `#[serde(flatten)]` on common.

```rust
#[derive(Debug, Serialize, Deserialize)]
struct Outer {
    key1: Option<String>,
    #[serde(flatten)]
    inner: Option<Inner>,
}

#[derive(Debug, Serialize, Deserialize)]
struct Inner {
    key2: String,
    key3: String,
}

fn main() {
    let mut m = HashMap::new();
    m.insert("key1", "v1");
    m.insert("key2", "v2");
    m.insert("key3", "v3");

    let v = serde_json::to_value(m).unwrap();
    let o: Outer = serde_json::from_value(v).unwrap();
    println!("{:?}", o);
}
```

Reply:
good catch. 👍